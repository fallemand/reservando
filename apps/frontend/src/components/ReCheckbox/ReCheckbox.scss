@import "~@/assets/styles/variables";

.re-checkbox {
  display: flex;
  position: relative;
  align-items: center;
  align-content: flex-start;
  color: $text-primary;

  &--full {
    flex-direction: row-reverse;
    background-color: transparent;
    transition: background-color 0.3s ease;
    border-radius: $bdu * 1.5;
    color: $border-primary;

    .re-checkbox-label {
      padding: 0.75em 1em;
      text-align: left;
      flex-grow: 1;
    }
    .re-checkbox-input {
      background-color: transparent;
      border-color: $border-secondary;
      border-radius: 50%;

      &:checked {
        & ~ .re-checkbox-label {
          color: $white;
          background-color: $border-primary;
        }
      }
    }
    .re-checkbox-check {
      $width-full: 0.875em;
      $height-full: 0.425em;
      left: calc(50% - #{$width-full/2});
      top: calc(50% - #{$height-full/2});
      height: $height-full;
      width: $width-full;
    }
    .re-checkbox-container {
      margin-right: 1em;
    }
  }

  &--checked {
    &.re-checkbox--full {
      background-color: $border-primary;
      color: $white;
    }
  }

  &-container {
    position: relative;
    display: flex;
    margin-right: 0.5em;
  }

  &-input {
    cursor: pointer;
    background-color: $white;
    height: 1em;
    width: 1em;
    font-size: 1em;
    padding: 0;
    // Add margin to the container, not here.
    // It will cause a misalignment in the check.
    margin: 0 !important;
    border: $border-primary 1px solid;
    -webkit-appearance: none;
    -moz-appearance: none;
    transition: background-color 0.2s ease-out, border-color 0.2s ease-out;
    border-radius: 2px;
  }

  &-label {
    cursor: pointer;
    align-self: stretch;
    font-size: 1em;
  }

  &-check {
    pointer-events: none;
    $width: 0.625em;
    $height: 0.25em;

    display: block;
    position: absolute;
    left: calc(50% - #{$width/2});
    top: calc(50% - #{$height/2});
    height: $height;
    width: $width;

    &::before {
      transition: border-color 0.3s ease;
      border-color: $border-primary;
      display: block;
      content: "";
      box-sizing: border-box;
      height: 100%;
      width: 100%;
    }
  }

  &-input:focus {
    outline: 0;
    border-color: $border-primary;
  }

  &-input:checked {
    background-color: $border-primary;
    border-color: $border-primary;

    & + .re-checkbox-check::before {
      transform: rotate(-45deg);
      border-left: 0.125em $white solid;
      border-bottom: 0.125em $white solid;
    }
  }

  &-input:indeterminate {
    background-color: $background-primary;
    border-color: $border-primary;

    & + .re-checkbox-check {
      $height: 0.125em;
      $width: 0.5em;
      left: calc(50% - #{$width/2});
      top: calc(50% - #{$height/2});
      height: $height;
      width: $width;
      border-bottom: $height $white solid;
    }
  }

  &-input:disabled {
    background: $surface-disabled;
    border-color: $border-primary;
    cursor: not-allowed;

    &:checked {
      background: $white;
    }

    & + .re-checkbox-check::before {
      border-color: $border-primary;
    }

    & ~ .re-checkbox-label {
      color: $text-secondary;
      cursor: not-allowed;
    }
  }
}
