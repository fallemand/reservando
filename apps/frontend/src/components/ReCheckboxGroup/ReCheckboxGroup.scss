.re-checkbox-group {
  display: flex;
  flex-direction: row;

  &--vertical {
    flex-direction: column;

    .re-checkbox:not(:last-child) {
      margin-bottom: 0.5em;
    }
  }
}
