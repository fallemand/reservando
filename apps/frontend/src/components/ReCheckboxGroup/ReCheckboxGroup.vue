<template>
  <div :class="`re-checkbox-group re-checkbox-group--${orientation}`">
    <!-- @slot Default slot for GygRadio components -->
    <slot />
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";

const ReCheckboxGroup = defineComponent({
  name: "ReCheckboxGroup",
  provide() {
    return {
      checkedItems: (id: string) => this.checked.includes(id),
    };
  },
  props: {
    orientation: { type: String, default: "vertical" },
    /** Id of item to show expanded. Array of ids if using multiple */
    checked: {
      type: Array as PropType<string[]>,
      required: true,
    },
  },
  methods: {
    handleItemChange(checked: boolean, id: string): void {
      let checkedItems: string[] | null = checked
        ? [...this.checked, id]
        : this.checked.filter((item) => item !== id);
      /** `(expandIds: string | string[])` */
      this.$emit("update:checked", checkedItems);
    },
  },
});

export default ReCheckboxGroup;
</script>

<style lang="scss" src="./ReCheckboxGroup.scss" />
