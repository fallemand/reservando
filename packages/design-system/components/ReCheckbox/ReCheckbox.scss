@import "../../styles/variables";

.re-checkbox {
  display: flex;
  position: relative;
  align-items: center;
  align-content: flex-start;
  color: $text-primary;
  transition: background-color 0.2s;
  padding: ($bdu * 1.5) $bdu;
  border-radius: $bdu;

  &:hover,
  &:active {
    background-color: $primary-100;
  }

  &--inverted {
    flex-direction: row-reverse;
  }

  &--contained {
    background-color: $white;
    border: 1px solid $grey-500;
    flex-direction: row-reverse;
  }

  &__container {
    position: relative;
    display: flex;
    margin-right: 0.5em;
  }

  &__input {
    cursor: pointer;
    background-color: $white;
    height: 1em;
    width: 1em;
    font-size: 1em;
    padding: 0;
    margin: 0 !important;
    border: $primary-300 2px solid;
    -webkit-appearance: none;
    -moz-appearance: none;
    transition: background-color 0.2s ease-out, border-color 0.2s ease-out;
    border-radius: $bdu / 2;

    &:focus {
      outline: 0;
    }
  }

  &__input:indeterminate {
    background-color: $border-primary;
    border-color: $border-primary;

    & + .re-checkbox__check {
      $height: 0.125em;
      $width: 0.5em;
      left: calc(50% - #{$width/2});
      top: calc(50% - #{$height/2});
      height: $height;
      width: $width;
      border-bottom: $height $white solid;

      &:before {
        transform: none;
      }
    }
  }

  &__label {
    cursor: pointer;
    align-self: stretch;
    font-size: 1em;
    flex-grow: 1;
    text-align: left;
  }

  &__check {
    pointer-events: none;
    $width: 0.625em;
    $height: 0.25em;

    display: block;
    position: absolute;
    left: calc(50% - #{$width/2});
    top: calc(50% - #{$height/2});
    height: $height;
    width: $width;

    &::before {
      transition: border-color 0.3s ease;
      border-color: $border-primary;
      display: block;
      content: "";
      box-sizing: border-box;
      height: 100%;
      width: 100%;
    }
  }

  &--checked {
    .re-checkbox__input {
      background-color: $primary-500;
      border-color: $primary-500;
    }

    .re-checkbox__check::before {
      transform: rotate(-45deg);
      border-left: 0.125em $white solid;
      border-bottom: 0.125em $white solid;
    }
  }

  &--disabled {
    .re-checkbox__input {
      background: $white;
      border-color: $grey-400;
      cursor: not-allowed;
    }

    .re-checkbox__check::before {
      border-color: $border-primary;
    }

    .re-checkbox__label {
      color: $text-secondary;
      cursor: not-allowed;
    }
  }
}
