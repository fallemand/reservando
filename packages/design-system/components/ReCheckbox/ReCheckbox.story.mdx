import { Meta, Story, ArgsTable, Canvas } from "@storybook/addon-docs/blocks";
import { ReCheckbox, ReCheckboxGroup } from "@reservando/design-system";

<Meta
  title="Components/Checkbox"
  component={ReCheckbox}
  argTypes={{
    modifier: { control: { type: "select", options: ["default", "full"] } },
  }}
/>

# ReCheckbox

<Canvas>
  <Story name="basic">
    {{
      components: { ReCheckbox },
      data() {
        return {
          checked: false,
        };
      },
      template: `
      <div>
        <ReCheckbox id="basic" v-model:checked="checked" label="Checkbox label" />
      </div>
    `,
    }}
  </Story>
</Canvas>

```html
<template>
  <ReCheckbox v-model:checked="isChecked" :id="my-checkbox" :label="Checkbox label" />
</template>

<script lang="ts">
  import { ReCheckbox } from "@reservando/design-system";
</script>
```

<ArgsTable components={{ ReCheckbox, ReCheckboxGroup }} />

#### Using modifiers

<Canvas>
  <Story name="modifiers">
    {{
      components: { ReCheckbox },
      data() {
        return {
          defaultChecked: false,
          inverted: false,
          contained: false,
        };
      },
      template: `
      <div style="max-width: 300px;">
        <ReCheckbox modifier="default" id="default" v-model:checked="defaultChecked" label="Default" />
        <ReCheckbox modifier="inverted" id="inverted" v-model:checked="inverted" label="Inverted" class="mb-1" />
        <ReCheckbox modifier="contained" id="contained" v-model:checked="contained" label="Contained" />
        <br />
        <ReCheckbox modifier="default" id="default-disabled" :disabled="true" v-model:checked="defaultChecked" label="Default disabled" />
        <ReCheckbox modifier="inverted" id="inverted-disabled" :disabled="true" v-model:checked="inverted" label="Inverted disabled" class="mb-1" />
        <ReCheckbox modifier="contained" id="contained-disabled" :disabled="true" v-model:checked="contained" label="Contained disabled" />
      </div>
    `,
    }}
  </Story>
</Canvas>

#### You can pass input default attributes

You can add the default attributes like `required` `disabled` `checked` on the element and they will be added to the input.

<Canvas>
  <Story name="attributes">
    {{
      components: { ReCheckbox },
      template: `
      <div>
        <ReCheckbox id="checked-1" checked="checked" label="Checked"/>
        <ReCheckbox id="indeterminate-1" checked="checked" :indeterminate="true" label="Indeterminate"/>
        <ReCheckbox id="disabled-1" disabled="disabled" label="Disabled"/>
        <ReCheckbox id="disabled-checked-1" checked="checked" disabled="disabled" label="Disabled checked"/>
      </div>
    `,
    }}
  </Story>
</Canvas>

## Using ReCheckboxGroup

<Canvas>
  <Story name="checkbox group">
    {{
      components: { ReCheckboxGroup, ReCheckbox },
      data() {
        return {
          itemsChecked: [],
        };
      },
      template: `
        <ReCheckboxGroup v-model:checked="itemsChecked">
          <ReCheckbox id="checkbox-1" label="Checkbox 1" />
          <ReCheckbox id="checkbox-2" label="Checkbox 2" />
          <ReCheckbox id="checkbox-3" label="Checkbox 3" />
        </ReCheckboxGroup>
        <p class="hint mt-4">Selected: {{ itemsChecked }}</p>
      `,
    }}
  </Story>
</Canvas>

```html
<template>
  <ReCheckboxGroup v-model:checked="itemsChecked">
    <ReCheckbox id="checkbox-1" label="Checkbox 1" />
    <ReCheckbox id="checkbox-2" label="Checkbox 2" />
    <ReCheckbox id="checkbox-3" label="Checkbox 3" />
  </ReCheckboxGroup>
</template>

<script lang="ts">
  import { ReCheckbox, ReCheckboxGroup } from "@reservando/design-system";
</script>
```
