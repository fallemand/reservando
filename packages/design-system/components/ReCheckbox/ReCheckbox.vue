<template>
  <div
    class="re-checkbox"
    :class="[`re-checkbox--${modifier}`, { 're-checkbox--checked': isChecked || indeterminate }]"
  >
    <div class="re-checkbox-container">
      <input
        :id="id"
        ref="input"
        type="checkbox"
        :checked="isChecked"
        class="re-checkbox-input"
        v-bind="$attrs"
        :indeterminate.prop="indeterminate"
        v-on="{
          change: emitChange,
        }"
      />
      <span class="re-checkbox-check" />
    </div>

    <label v-if="label || $slots.label" class="re-checkbox-label" :for="id">
      <!-- @slot Useful to pass HTML or something custom -->
      <slot name="label-slot">
        {{ label }}
      </slot>
    </label>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from "vue";
export const MODIFIERS = ["default", "full"];

const ReCheckbox = defineComponent({
  name: "ReCheckbox",
  inject: ["checkedItems"],
  props: {
    /** Needed for the attribute `for`. Should be unique on the html doc level */
    id: { type: String, required: true },
    label: { type: String, default: null },
    /** Sets `indeterminate` value, which puts checkbox in "maybe" state */
    indeterminate: { type: Boolean, default: false },
    modifier: {
      default: "default",
      type: String as PropType<typeof MODIFIERS[number]>,
      validator: (value: string): boolean => MODIFIERS.includes(value),
    },
  },
  computed: {
    isChecked(): boolean {
      return this.checkedItems ? this.checkedItems(this.id) : this.$attrs.checked;
    },
  },
  methods: {
    emitChange($evt: Event): void {
      const checked = ($evt.target as HTMLInputElement).checked;
      if (this.$parent && this.$parent.$options.name === "ReCheckboxGroup") {
        const checkboxGroup = this.$parent;
        checkboxGroup.handleItemChange(checked, this.id);
      } else {
        /** `(value: string, $evt: Event)`<br /> fired when changed */
        this.$emit("update:checked", checked, $evt);
      }
    },
  },
});

export default ReCheckbox;
</script>

<style lang="scss" src="./ReCheckbox.scss" />
