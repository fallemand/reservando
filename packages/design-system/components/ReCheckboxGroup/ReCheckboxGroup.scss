.re-checkbox-group {
  display: flex;
  flex-direction: row;

  &--vertical {
    flex-direction: column;
  }
}
