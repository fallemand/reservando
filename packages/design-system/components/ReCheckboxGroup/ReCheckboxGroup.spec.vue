<script lang="ts">
import { shallowMount, mount } from "@vue/test-utils";
import ReCheckboxGroup from "./ReCheckboxGroup.vue";
import ReCheckbox from "../ReCheckbox/ReCheckbox.vue";

describe("ReCheckboxGroup component", () => {
  it("should render with default props", () => {
    const wrapper = shallowMount(ReCheckboxGroup, {
      props: {
        checked: [],
      },
      slots: {
        default: "__ReCheckbox-Items__",
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should call @update:checked on value change", () => {
    const wrapper = shallowMount(ReCheckboxGroup, {
      props: {
        checked: [],
      },
    });
    // Check id-1
    wrapper.vm.handleItemChange(true, "id-1");
    expect(wrapper.emitted()["update:checked"][0]).toEqual([["id-1"]]);
    // Check id-2
    wrapper.vm.handleItemChange(true, "id-2");
    expect(wrapper.emitted()["update:checked"][1]).toEqual([["id-1", "id-2"]]);
    // Uncheck id-2
    wrapper.vm.handleItemChange(false, "id-2");
    expect(wrapper.emitted()["update:checked"][2]).toEqual([["id-1"]]);
  });
});
</script>
