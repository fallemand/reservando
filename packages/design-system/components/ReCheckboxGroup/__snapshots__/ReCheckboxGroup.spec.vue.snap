// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReCheckboxGroup component should render with default props 1`] = `
<div
  class="re-checkbox-group re-checkbox-group--vertical"
>
  <!-- @slot Default slot for ReCheckbox components -->
  
  __ReCheckbox-Items__
  
</div>
`;
